### DISCUSSION TITLE
is it possible to load a client-side GLB blob into model-viewer?

### DISCUSSION THREAD

**Discussion by rawwerks** [google/model-viewer#3184] [Q&A]
@njiang and i are running into some challenges taking a .glb file that was generated client-side and passing that to model-viewer as a blob. 

this post suggests that scene-viewer cannot accept blobs https://github.com/google/model-viewer/discussions/2884 - but before we go build a bunch of webXR stuff from scratch  - i'm curious if someone here (maybe @elalish ?) could help give a definitive answer. 

what we are trying to accomplish is: 
- the user uploads a .stl file
- there is client-side js code that converts that to a GLTF and packs it to GLB
- (at least for Android devices) - this GLB blob is then passed to model-viewer to be viewed in AR, ideally without ever getting sent to our server

we currently have this working w/ three.js and a USDZ export for iOS AR Quick Look. we are thinking that model-viewer might be the best way to also support Android devices.

**Comment by samaneh-kazemi** [comment]
Hi @rawwerks, thanks for reaching out. Unfortunately, as you saw in the other discussion #2884, passing GLB blob to src in SceneViewer mode is not supported. This is due to security restrictions in Android. Basically because scene viewer is technically a separate app(though a 1P app), the blob is meaningless outside of Chrome. Quick Look only works because Apple makes an exception for it, but tighter securities are applied in Android even between 1P apps.

**Reply by rawwerks** [reply]
Ok, that is unfortunate. Thank you for clarifying @samaneh-kazemi.

So what is the best practice for making this as seamless as possible for the user? Upload blob to server and have the user re-download from URL? I'd like to avoid the QR code => piping server workflow that model-viewer employs...I just think we're going to lose a lot of users at that step.

**Reply by elalish** [reply]
Well, it's only Scene Viewer that's blocked from accepting blobs; WebXR (our default AR mode) works just fine. Have you tried this?

**Reply by rawwerks** [reply]
Thank you @elalish - this is exactly what I was hoping to clarify before we built some crazy workaround. 

So just to be really sure I am understanding correctly - if I generate my GLTF 2.0 GLB client side and feed that as a blob to model-viewer, then model-viewer should be able to automatically take care of:
- in browser preview via three.js / webGL 
- running usdzexporter.js (from three.js) and blobbing USDZ for Quick Look when iOS detected
- running AR via WebXR for Android devices

Is this correct?

If this is true, I can probably rewrite my current application with 1% of the current lines of code. I just don't want to set out on this refactoring mission without the confidence that it is possible. (Which it seems it should be ok to blob => model-viewer, given that the model-viewer editor runs everything except piping locally.)

**Reply by elalish** [reply]
Yes, I believe that is all true. If you open our editor on your phone and load a model from your file system this is effectively what happens. It shouldn't be hard to give it a quick test before you do a major refactor, but yes, our goal is to make this kind of thing easy for you.

**Reply by DanieleMagri** [reply]
Hi! I did a 3D configurator of a product using Three.js and i didn't have any issues making the object viewable in AR on Apple devices through QuickLook, i use a blob from the USDZ exporter and the new 3D model is avaiable on the go, ready to be viewed, with all the customizations done.
My client asked me to make this possible on Android as well, so I decided to use model-viewer, which from what I read should also automatically take care of the conversion to USDZ for QuickLook, I also discovered that scene-viewer (the native android solution similar to QuickLook) doesn't support blobs, so i've decided to use WebXR through model-viewer.

But I have 2 problems that I can't solve:

1- i can't pass the blob to model-viewer. How you do it? Apparently the same method I used for QuickLook doesn't seem to work, which is this (i renamed the model "3Dmodel" for you to understand):
![immagine](https://user-images.githubusercontent.com/60880674/169835871-4a005926-d7f7-4fa5-8da9-b805d854a53b.png)
![immagine](https://user-images.githubusercontent.com/60880674/169836294-42fe9524-c38e-4a5d-be16-1ec18a3b4432.png)

2- To start the AR scene, you press the model-viewer button, which is on the same page where the 3D scene of the configurator created in Three.js is present, the problem is that when I start the AR view, the camera opens for some time (4 frame) and after that everything crashes and returns to the configurator page, with the 3D configurator scene "broken", the object is no longer there.
I also tried to insert model-viewer in another page of the web app without the configurator, and if started from there, it doesn't give any problem, everything works, but if I enter in the configurator page, and then I return to the page without the configurator and start the AR display, it works, but it lags, so I thought the problem is the configurator, since WebXR is also based on Three.js, and I noticed that exiting the configurator, the animate () function of the configurator continues to work, and I think that is the problem.

**Reply by elalish** [reply]
For 1), first off, you might not need three.js at all; if you're only configuring materials, that can happen through MV's public API the same way our editor does. If you do need to use three.js, you can either use `GLTFExporter` and set that blob as our `src` (we only support glTF/GLB input, not USDZ), or you can hack down into our `$scene` private symbol, which is a three.js scene and then you can do whatever you want to it. Just be careful to test before upgrading versions, as we don't promise stability in the non-public API. 

For 2) I think you're on the right track. MV is there to handle the three.js internals, so if you're doing much of your own they are probably in conflict.

**Reply by DanieleMagri** [reply]
First of all, thanks for the reply :)

For 2) I had two scenes, and two renderers (the configurator one and the MV one) so I thought I'd do something that turned out to be the right choice: I deleted the animate () function and the renderer of my configurator, leaving only the "backend" (the part that deals with loading the model and changing textures), **without** displaying the model and the configurator canvas to be clear, and using MV instead to display the model, seems to work well, even the AR now does not give any problems.
In summary:
Three.js changes a texture, calls the GLTFExporter function which changes the value of src by first setting it to null and then linking the blob with the updated model which is then rendered via MV.

For 1) For this configurator I use specific textures saved locally (/assets/textures), would it still be possible to change them using MV's public API?
Last questions:
would it be possible to use a GLB model directly and use it in model-viewer without going through the converter?
i didn't tried quicklook with MV yet, it should automatically convert the model to USDZ right?

**Reply by elalish** [reply]
Yes to all of these questions. Have a look at these examples for texture swapping: https://modelviewer.dev/examples/scenegraph/#createTexturesExample

The whole point of MV is to use a GLB directly. Our USDZ auto-conversion is detailed here: https://modelviewer.dev/docs/#entrydocs-augmentedreality-attributes-arModes

**Reply by DanieleMagri** [reply]
Very interesting, I took a look at it, but what if I had a 3D model part that I wanted to change? For example: I have a model of a staircase in glb with different meshes for the handrails integrated in the model, would it be possible through a selection to turn on the selected handrail mesh and turn off the others? (this is the method I currently use on three.js within the configurator)

[Any updates?](https://github.com/google/model-viewer/issues/1774#issuecomment-838742368)

**Reply by elalish** [reply]
I'm actually looking at building an API specifically for this right now, but in the meanwhile, yes, you should be able to achieve it by having a separate material for each mesh and making them alternately transparent or not.
